dlt/destinations/synapse/sql_client.py

```diff
@@ -0,0 +1,342 @@
+
+    def drop_dataset(self) -> None:
+        rows = self.execute_sql(
+            "SELECT table_name FROM information_schema.views WHERE table_schema = %s;", self.dataset_name
+        )
+        view_names = [row[0] for row in rows]
+        self._drop_views(*view_names)
+        rows = self.execute_sql(
+            "SELECT table_name FROM information_schema.tables WHERE table_schema = %s;", self.dataset_name
+        )
+        table_names = [row[0] for row in rows]
+        self.drop_tables(*table_names)
+
+        self.execute_sql("DROP SCHEMA %s;" % self.fully_qualified_dataset_name())
+
+    def table_exists(self, table_name: str) -> bool:
+        query = """
+        SELECT COUNT(*)
+        FROM INFORMATION_SCHEMA.TABLES
+        WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
+        """
+        result = self.execute_sql(query, self.dataset_name, table_name)
+        return result[0][0] > 0
+
+    def drop_tables(self, *tables: str) -> None:
+        if not tables:
+            return
+        for table in tables:
+            if self.table_exists(table):
+                self.execute_sql(f"DROP TABLE {self.make_qualified_table_name(table)};")
+
+    def _drop_views(self, *tables: str) -> None:
+        if not tables:
+            return
+        statements = [f"DROP VIEW {self.make_qualified_table_name(table)};" for table in tables]
+        self.execute_fragments(statements)
+
+    import pyodbc
+
+    def set_input_sizes(self, *args):
+        if len(args) == 1 and isinstance(args[0], tuple):
+            args = args[0]
+
+        args = list(args)  # Convert tuple to list for modification
+        input_sizes = []
+        for index, arg in enumerate(args):
+            if isinstance(arg, str):
+                # Check if the string is a valid JSON
+                if is_valid_json(arg):
+                    input_sizes.append((pyodbc.SQL_WVARCHAR, 0, 0))  # Set input size for JSON columns
+                else:
+                    # Assuming ntext is represented as a regular string
+                    # You may want to have a more robust check for ntext data here
+                    input_sizes.append((pyodbc.SQL_WVARCHAR, 0, 0))  # Set input size for ntext columns
+            else:
+                input_sizes.append(None)  # Default handling for non-string types
+
+        logger.debug(f"Input sizes: {input_sizes}")
+        return input_sizes, tuple(args)
+
+    def execute_sql(self, sql: AnyStr, *args: Any, **kwargs: Any) -> Optional[Sequence[Sequence[Any]]]:
+        #logging.info(f"Executing SQL: {sql} with arguments: {args} and keyword arguments: {kwargs}")
+        # Convert any JSON arguments to strings
+        args = tuple(json.dumps(arg) if isinstance(arg, dict) else arg for arg in args)
+
+        param_values = args  # This line remains unchanged as you're capturing all additional arguments into param_values
+        original_sql = sql
+
+        # Check if it's a multi-row insert
+        is_multi_row_insert = (
```

Comment:
you should implement a custom InsertValuesLoadJob that does this and not hack the execute_sql function.

Reply:
Thank you for the feedback, @sh-rp.  How about a subclass of InsertValuesLoadJob tailored for Microsoft Synapse that lives in synapse.py?

Given that DLT handles multiple destinations and insert_job_client.py provides a foundation for all of them, I can isolate Synapse-specific conversions and avoid execute_sql

SynapseInsertValuesLoadJob

Reply:
@sh-rp circling back to this, please let me know if this is in right direction.  generate_insert_query() now lives in synapse.py 

Instead of taking a SQL string (vulnerable to injection) it now takes comparable params to InsertValuesLoadJob